client/src/app/containers/Game/useGame.ts

```diff
@@ -0,0 +1,235 @@
+import axios from 'axios';
+import { useDispatch, useSelector } from 'react-redux';
+
+import { useAuth } from '../../../auth/useAuth';
+import { RootState } from '../../../store/store';
+import { useAlert } from '../Alert/useAlert';
+import { Card, Game, Phase } from './models';
+import {
+    actions,
+    CountCribAction,
+    CountHandAction,
+    CribAction,
+    CutAction,
+    DealAction,
+    GameAction,
+    PegAction,
+} from './slice';
+import { CreateGameResponse } from './types';
+
+interface Result {
+    game: Game;
+    selectedCards: Card[];
+    createGame: (playerIDs: string[]) => Promise<void>;
+    getGame: (id: number) => Promise<void>;
+    refreshGame: () => Promise<void>;
+    toggleSelectedCard: (c: Card) => void;
+    clearSelectedCards: () => void;
+    submitDealAction: (a: DealAction) => Promise<void>;
+    submitBuildCribAction: (a: CribAction) => Promise<void>;
+    submitCutDeckAction: (a: CutAction) => Promise<void>;
+    submitPegAction: (a: PegAction) => Promise<void>;
+    submitCountHandAction: (a: CountHandAction) => Promise<void>;
+    submitCountCribAction: (a: CountCribAction) => Promise<void>;
+}
+// getPlayerAction returns the JSON struct which the server knows
+// how to interpret
+interface ServerCard {
+    s: number;
+    v: number;
+}
+interface ActionRequest {
+    pID: string;
+    gID: number;
+    o: ReturnType<typeof mapPhaseToOverComes>;
+    a: {
+        ns?: number;
+        cs?: ServerCard[];
+        p?: number;
+        sg?: boolean;
+        c?: ServerCard;
+        pts?: number;
+    };
+}
+
+const mapPhaseToOverComes = (p: Phase) => {
+    if (p === 'Deal') {
+        return 0;
+    }
+    if (p === 'BuildCrib') {
+        return 1;
+    }
+    if (p === 'Cut') {
+        return 2;
+    }
+    if (p === 'Pegging') {
+        return 3;
+    }
+    if (p === 'Counting') {
+        return 4;
+    }
+    if (p === 'CribCounting') {
+        return 5;
+    }
+    return -1;
+};
+
+const cardToGolangCard = (c: Card): ServerCard => {
+    const magicMap = {
+        Spades: 0,
+        Clubs: 1,
+        Diamonds: 2,
+        Hearts: 3,
+    };
+    return {
+        s: magicMap[c.suit],
+        v: c.value,
+    };
+};
+
+function getPlayerAction<T extends GameAction>(
+    myID: string,
+    gID: number,
+    phase: Phase,
+    currentAction: T,
+): ActionRequest {
+    const request: ActionRequest = {
+        pID: myID,
+        gID,
+        o: mapPhaseToOverComes(phase),
+        a: {},
+    };
+    switch (phase) {
+        case 'Deal':
+            request.a = { ns: (currentAction as DealAction).numShuffles };
```

Comment:
:mag: :bulb: instead of switching on the `phase`, should we instead switch on the type of `currentAction`? and should we call `currentAction` something like `submittedAction` or something?

Reply:
I'd love to! Except TypeScript doesn't have runtime type checking; only compile time type checking, so I don't think this is possible (unless we explicitly set an `ActionKind` or something, which I'm not opposed to)

Reply:
Let's keep it as is. That's why I built it this way; to allow for dynamically sending a player action and then unmarshalling it into multiple golang structs.